### PR TITLE
[FAB-17619] encapsulate config proto in new type

### PR DESCRIPTION
**Pull request by stephyee**
#### Type of change

- Improvement (improvement to code, performance, etc)

#### Description

Wrap config transaction proto and move API to wrapped type

#### Related issues
[FAB-17619](https://jira.hyperledger.org/browse/FAB-17619)

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -60,6 +60,14 @@ type standardConfigPolicy struct {
 	value *cb.Policy
 }
 
+// Config is the proto configuration of a channel
```
Comment (stephyee) [inline pkg/config/config.go]: wraps a config transaction

File: pkg/config/example_test.go [hunk 2]
```diff
@@ -460,6 +460,11 @@ func Example_usage() {
 	// Clone the base config for processing updates
 	updatedConfig := proto.Clone(baseConfig).(*cb.Config)
 
+	c := config.Config{
```
Comment (stephyee) [inline pkg/config/example_test.go]: Add a constructor for creating config...

File: pkg/config/config.go (hunk 1 above)
Comment (stephyee) [reply pkg/config/config.go]: an idea `c := configtx.New(config)`

File: pkg/config/msp.go [hunk 3]
```diff
@@ -186,19 +186,100 @@ func GetMSPConfigurationForConsortiumOrg(config *cb.Config, consortiumName, orgN
...
+	return nil
+}
+
+// YEAR is a time duration for a standard 365 day year.
+const YEAR = 365 * 24 * time.Hour
+
+// RevokeCertificateFromMSP takes a variadic list of x509 certificates, creates
+// a new CRL signed by the specified ca certificate and private key, and appends
+// it to the revocation list for the specified application org MSP.
+func (c *Config) RevokeCertificateFromMSP(orgName string, caCert *x509.Certificate, caPrivKey *ecdsa.PrivateKey, certs ...*x509.Certificate) error {
```
Comment (sykesm) [inline pkg/config/msp.go]: A few things

- I missed the singular/plural of 'RevokeCertificate' - it's variadic so certificate should be plural
- The name implies something more imperative than I believe it does. This really creates and adds a CRL.
- This is another place where a mutable MSP would be conceptually simpler to deal with

Even though this is exactly what the story asked for, stepping back, I wonder if we just want to provide a way to add a CRL and a function that creates a CRL with a SigningIdentity.

**Review by sykesm — COMMENTED** [review]
Looks like a fairly straightforward and mechanical change. No major concerns there.

- The code does / will have merge conflicts that need to be addressed.
- The example with state is looking better but highlights how a "chaining" API or an API with mutable intermediate objects may be easier to consume since most of the methods have multiple qualifiers in their names.
- Looks like most getters can be converted to drop the Get prefix.

**Comment by stephyee** [comment]
I forgot to fix the int test, working on that.

File: pkg/config/msp.go (hunk 3 above)
Comment (wlahti) [reply pkg/config/msp.go]: I'll create a story for this MSP-related work.

File: pkg/config/msp.go [hunk 4]
```diff
@@ -153,30 +153,30 @@ type NodeOUs struct {
 
 // GetMSPConfigurationForApplicationOrg returns the MSP configuration for an existing application
 // org in a config transaction.
-func GetMSPConfigurationForApplicationOrg(config *cb.Config, orgName string) (MSP, error) {
-	applicationOrgGroup, ok := config.ChannelGroup.Groups[ApplicationGroupKey].Groups[orgName]
+func (c *Config) GetMSPConfigurationForApplicationOrg(orgName string) (MSP, error) {
```
[the remark this answers is not in the document]
Comment (wlahti) [reply pkg/config/msp.go]: I'll incorporate this into the story I'm creating.

File: pkg/config/msp.go [hunk 5]
```diff
@@ -186,19 +186,100 @@ func GetMSPConfigurationForConsortiumOrg(config *cb.Config, consortiumName, orgN
...
 	}
 
-	return getMSPConfigForOrg(consortiumOrgGroup, orgName)
+	return getMSPConfig(consortiumOrgGroup)
 }
 
-// getMSPConfigForOrg parses the MSP value in a config group for a specific org and returns
+// AddRootCAToMSP takes a root CA x509 certificate and adds it to the
+// list of rootCerts for the specified application org MSP.
+func (c *Config) AddRootCAToMSP(rootCA *x509.Certificate, orgName string) error {
```
[the remark this answers is not in the document]
Comment (wlahti) [reply pkg/config/msp.go]: I'll also incorporate this.

File: integration/config/config_test.go [hunk 6]
```diff
@@ -147,13 +147,18 @@ var _ = Describe("Config", func() {
 			channelConfig := nwo.GetConfig(network, peer, orderer, "testchannel")
 			updatedChannelConfig := proto.Clone(channelConfig).(*cb.Config)
 
+			c := config.Config{
+				Base:    channelConfig,
+				Updated: updatedChannelConfig,
```
[the remark this answers is not in the document]
Comment (stephyee) [reply integration/config/config_test.go]: 🛠

File: pkg/config/config.go [hunk 7]
```diff
@@ -60,6 +60,23 @@ type standardConfigPolicy struct {
 	value *cb.Policy
 }
 
+// Config wraps a config transaction
+type Config struct {
+	// original state of the config
+	Base *cb.Config
+	// modified state of the config
+	Updated *cb.Config
```
[the remark this answers is not in the document]
Comment (stephyee) [reply pkg/config/config.go]: 🛠

File: pkg/config/application.go [hunk 8]
```diff
@@ -111,8 +111,8 @@ func RemoveAnchorPeer(config *cb.Config, orgName string, anchorPeerToRemove Anch
 }
 
 // GetAnchorPeers retrieves existing anchor peers from a application organization.
-func GetAnchorPeers(config *cb.Config, orgName string) ([]AnchorPeer, error) {
-	applicationOrgGroup, ok := config.ChannelGroup.Groups[ApplicationGroupKey].Groups[orgName]
+func (c *Config) GetAnchorPeers(orgName string) ([]AnchorPeer, error) {
```
[the remark this answers is not in the document]
Comment (stephyee) [reply pkg/config/application.go]: I fix it a follow up commit.

File: pkg/config/msp.go [hunk 9]
```diff
@@ -700,9 +781,6 @@ func getOrgMSPValue(org *cb.ConfigGroup) (*cb.ConfigValue, error) {
 	return configValue, nil
 }
 
-// YEAR is a time duration for a standard 365 day year.
-const YEAR = 365 * 24 * time.Hour
-
 // RevokeCertificateFromMSP takes a variadic list of x509 certificates, creates
```
Comment (wlahti) [inline pkg/config/msp.go]: Do we want to leave around this old version? Same question for AddRootCAToMSP.

Also, the new version (aka the one on ConfigTx) doesn't have the latest code for this function. Notably, it's missing the new validateCert logic.

**Review by wlahti — CHANGES_REQUESTED** [review]
The RevokeCertificate function on ConfigTx needs to be updated to incorporate the recently merged code for validating certs.

File: pkg/config/msp.go (hunk 9 above)
Comment (stephyee) [reply pkg/config/msp.go]: I think I left this after resolving some merge conflicts. I’ll fix it.

Comment (stephyee) [reply pkg/config/msp.go]: 🛠

**Review by wlahti — APPROVED** [review]
Thanks! Looks good to me now!

**Review by sykesm — APPROVED** [review]

**Pull request merged**